boot/LoadReactOS.c

```diff
@@ -0,0 +1,410 @@
+	 * spec */
+
+	"movw $0x0018, %ax \n"
+	"mov %eax, %ss \n"
+	"mov %eax, %ds \n"
+	"mov %eax, %es \n"
+	"mov %eax, %fs \n"
+	"mov %eax, %gs \n"
+
+	/* Set the stack pointer to give us a valid stack */
+	"movl $0x03BFFFFC, %esp \n"
+	);
+
+	/* Ok, preparations are complete. Now move the image.
+	 * Values of load_end_addr and bss_end_addr can be equal to zero, see:
+	 * https://www.gnu.org/software/grub/manual/multiboot/multiboot.html#Header-address-fields */
+	memcpy((u8 *)mbHeader->load_addr,
+		(u8 *)mbHeader - (mbHeader->header_addr - mbHeader->load_addr),
+		(mbHeader->load_end_addr > 0 ? mbHeader->load_end_addr - mbHeader->load_addr : ldrSize));
+	if (mbHeader->bss_end_addr > 0)
+		memset((u8 *)mbHeader->load_end_addr, 0x00, mbHeader->bss_end_addr - mbHeader->load_end_addr);
+
+	/* Set up the multiboot info structure */
+	infoPtr = mbHeader->bss_end_addr;
+	if (infoPtr == 0)
+		infoPtr = mbHeader->load_addr + ldrSize;
+	mbInfo = (PMULTIBOOTINFO)((infoPtr + 3) & ~0x3);
+	memset(mbInfo, 0x00, sizeof(MULTIBOOTINFO));
+	mbInfo->flags = MB_INFO_FLAG_MEM_SIZE | MB_INFO_FLAG_BOOT_DEVICE | MB_INFO_FLAG_MEMORY_MAP;
+	/* Multiboot spec says mem_lower can't be larger than 640, which is
+	 * true for a PC. Since we're bending the multiboot rules anyway
+	 * let's pass the full megabyte */
+	mbInfo->mem_lower = 1024;
+	mbInfo->mem_upper = (xbox_ram - 1) * 1024;
+	mbInfo->boot_device = bootDev;
+	mbInfo->mmap_length = 2 * sizeof(MEMORYMAP);
+	mmap = (PMEMORYMAP)(mbInfo + 1);
+	mbInfo->mmap_addr = (u32)mmap + sizeof(u32);
+	/* Normal RAM */
+	mmap->size = sizeof(MEMORYMAP);
+	mmap->base_addr_low = 0;
+	mmap->base_addr_high = 0;
+	mmap->length_low = (xbox_ram * 1024 * 1024) - FB_SIZE;
+	mmap->length_high = 0;
+	mmap->type = 1;
+	/* Video RAM */
+	mmap++;
+	mmap->size = sizeof(MEMORYMAP);
+	/* FIXME: Framebuffer is initialized statically with FB_START in BootVgaInitialization.c
+	 * Remove this condition once the problem is fixed */
+	if (xbox_ram > 64) {
+		mmap->base_addr_low = (64 * 1024 * 1024) - FB_SIZE;
+	} else {
+		mmap->base_addr_low = (xbox_ram * 1024 * 1024) - FB_SIZE;
+	}
+	mmap->base_addr_high = 0;
+	mmap->length_low = FB_SIZE;
+	mmap->length_high = 0;
+	mmap->type = 0;
+	if (xbox_ram > 64) {
+		/* FIXME: Remove this block once framebuffer problem is fixed */
+		mmap++;
+		mmap->size = sizeof(MEMORYMAP);
+		mmap->base_addr_low = 64 * 1024 * 1024;
+		mmap->base_addr_high = 0;
+		mmap->length_low = (xbox_ram - 64) * 1024 * 1024;
+		mmap->length_high = 0;
+		mmap->type = 1;
+	}
+
+	/* Now setup the registers and jump to the multiboot entry point */
+	asm volatile (
+	"xorl %%ecx,%%ecx \n"
+	"xorl %%edx,%%edx \n"
+	"xorl %%esi,%%esi \n"
+	"xorl %%edi,%%edi \n"
+	"pushl %%eax \n"
+	"movl %0,%%eax \n"
+	"ret \n"
+	: : "i" (MULTIBOOT_BOOTLOADER_MAGIC), "b" (mbInfo), "a" (mbHeader->entry_addr));
+
+	// We are not longer here, we are already in freeldr, we never come back here
+
+	// See you again in ReactOS then
+	while (1);
+}
```

Comment:
This entire file (and commit) feels like a misnomer. This looks like it is about adding support for multiboot, not ReactOS. ReactOS support only appears to be the positive sideeffect and primary motivation of having multiboot. Or is multiboot exclusive to ReactOS (by design, or in this implementation)?

If this code is not exclusive to ReactOS, then it shouldn't refer to ReactOS explicitly.

Reply:
Yes it's not exclusive to ReactOS. But right now it's tied to the freeldr bootloader, and would require more refactoring to have universal multiboot support.

Reply:
Fair enough - I don't care enough for Cromwell to argue this.
Ideally we'd probably have this as `LoadMultiboot` or something, and then only some ReactOS specific functions for finding it, however, we might never need this and we can always just rename / refactor later.